use tmux passthrough for kitty graphics protocol

l4zygreed opened the pull request:
Updated Ethan PR for current master

<!-- Provide a descriptive summary of the changes in the title above -->

#### ISSUE TYPE
<!-- Pick relevant types and delete the rest -->
- Improvement/feature implementation

#### RUNTIME ENVIRONMENT
<!-- Details of your runtime environment -->
<!-- Retrieve Python/ranger version and locale with `ranger -\-version` -->
- Operating system and version: Macos sequoia 15.4
- Terminal emulator and version: Ghostty 1.1.3
- Python version: 3.13.3
- Ranger version/commit: ranger-master v1.9.3-794-gcf51b12f
- Locale: en_US.UTF-8

#### CHECKLIST
<!-- All [REQUIRED] requisites need to be fulfilled -->
<!-- Replace [ ] with [X] when fulfilled -->
- [x] The `CONTRIBUTING` document has been read **[REQUIRED]**
- [x] All changes follow the code style **[REQUIRED]**
- [x] All new and existing tests pass **[REQUIRED]**
- [x] Changes require config files to be updated
    - [x] Config files have been updated

#### DESCRIPTION
<!-- Describe the changes in detail -->
Updated PR #2856 to be compatible with current master branch.

#### MOTIVATION AND CONTEXT
Want to finally see images inside tmux on my mac and outside of iterm/kitty terminal 😆 
Found old Ethan PR and decided to resolve conflicts and update it for current master 

#### TESTING

- `make test` executed successfully 
- kitty without tmux
- kitty with tmux and allow-passthrough not set
- kitty with tmux and allow-passthrough set
- ghostty without tmux
- ghostty with tmux and allow-passthrough not set
- ghostty with tmux and allow-passthrough set

#### IMAGES / VIDEOS<!-- Only if relevant -->
<!-- Link or embed images and videos of screenshots, sketches etc. -->

https://github.com/user-attachments/assets/a550437e-80e7-48b0-ade5-fe03c7ff761d

Ethsan commented on ranger/ext/img_display.py:
```diff
@@ -805,6 +844,8 @@ def draw(self, path, start_x, start_y, width, height):
         if self.needs_late_init:
             self._late_init()
+        if self.use_placeholder:
+            cmds.update({'U': 1})
```
```suggestion
            # U: 1=enable virtual placement using unicode placeholder
            cmds.update({'U': 1})
```
The past me didn't seems to like comment

Ethsan commented on ranger/ext/img_display.py:
```diff
@@ -738,16 +759,17 @@ def _late_init(self):
             resp = b''
             #          (DA1 resp start   )     (DA1 resp end     )
-            while not ((b'\x1b[?' in resp) and (resp[-1:] == b'c')):
+            # while not ((b'\x1b[?' in resp) and (resp[-2:] == self.answer_end)):
```
remove the comment

Ethsan commented on ranger/ext/img_display.py:
```diff
@@ -694,6 +706,10 @@ class KittyImageDisplayer(ImageDisplayer, FileManagerAware):
         https://github.com/kovidgoyal/kitty/blob/master/graphics-protocol.asciidoc"""
```
link seem broken, we can maybe update it in a separate commit ? With like https://sw.kovidgoyal.net/kitty/graphics-protocol/#display-images-on-screen ?

Ethsan reviewed (commented):
Hey, thanks for the keeping the pr alive and the update, it's cleaner now!

Should we separate the commit in two ? Something like  :
- support for virtual placement using unicode placeholder
- support for tmux allow-passthrough mode

Moreover some minor changes below.

Ethsan commented on ranger/ext/img_display.py:
```diff
@@ -43,6 +44,16 @@
+# | values are specified with diacritics listed in [rowcolumn-diacritics.txt]
+# Below is the list of the diatrics
+# pylint: disable-next=line-too-long
+KITTY_DIACRITICS = ["\U00000305", "\U0000030D", "\U0000030E", "\U00000310", "\U00000312", "\U0000033D", "\U0000033E", "\U0000033F", "\U00000346", "\U0000034A", "\U0000034B", "\U0000034C", "\U00000350", "\U00000351", "\U00000352", "\U00000357", "\U0000035B", "\U00000363", "\U00000364", "\U00000365", "\U00000366", "\U00000367", "\U00000368", "\U00000369", "\U0000036A", "\U0000036B", "\U0000036C", " ...
```
I am still ashamed of this lines, I am open to better idea. Maybe putting the list inside a separate python file ? or json file ?

Ethsan replied:
Hi,

> I think parsing json each time is not a good idea. It’s not a sensative/dynamic data or config to put it into json or another format which will require additional resources and time for parsing.

Agreed

> 
> It’ll probably will be instantaneous on most modern systems, but anyway, why do more when we can do less.
> 
> About moving to another python file, idk seems fine to me. I think this is for maintainers to decide.

@toonn, what do you prefer ?

Ethsan replied:
I don't think the unicode code points will change, since it would break all current implementations of the kitty graphical protocol.

l4zygreed replied:
For me having it there is totally fine. And creating a separate file for a single list sounds meh… 😐

And also don’t think that they'll change unicode points after 2+ years after they added 🤔 

So I guess we can mark this as resolved?

toonn replied:
I'd still prefer an upstream source or at least a copy of an upstream source so that it's easy to verify it's still in sync down the line.

Any reason not to, past "It's every so slightly easier not to read bytes from a file?"

l4zygreed committed:
support for virtual placement using unicode placeholder

Updated Ethan PR for current master

Co-authored-by: Ethan <milonthan@gmail.com>

l4zygreed committed:
support for tmux allow-passthrough mode

Updated Ethan PR for current master

Co-authored-by: Ethan <milonthan@gmail.com>

Ethsan commented:
Resolve #2414, Resolve #2846, Resolve #3072

Ethsan committed:
fixup! support for virtual placement using unicode placeholder

Ethsan commented:
Hi,
@toonn do you prefer something like that ? or should we revert ?

l4zygreed commented on ranger/ext/img_display.py:
```diff
@@ -898,6 +934,57 @@ def _format_cmd_str(self, cmd, payload=None, max_slice_len=2048):
+
+        try:
+            with open(
+                self.fm.relpath("rowcolumn-diacritics.txt"),
```
```suggestion
                self.fm.relpath(“data”, “rowcolumn-diacritics.txt"),
```
@Ethsan you placed file in data folder, so the path is incorrect and throws error.
also why file named txt while it’s treated like csv? 🤔

Ethsan replied:
Hi,
> @Ethsan you placed file in data folder, so the path is incorrect and throws error.

Yeah sorry about that, I move it will trying ot fix linting issues and forgot to update path.

> also why file named txt while it’s treated like csv? 🤔

I have simply reused the name given upstream [rowcolumn-diacritics.txt](https://github.com/kovidgoyal/kitty/blob/master/gen/rowcolumn-diacritics.txt)

toonn commented on ranger/data/rowcolumn-diacritics.txt:
I was kinda hoping to read this from wherever Kitty or Ghostty install it on the system but I guess this makes it compatible with other terminals too.

toonn commented on ranger/ext/img_display.py:
```diff
@@ -898,6 +934,57 @@ def _format_cmd_str(self, cmd, payload=None, max_slice_len=2048):
+                newline="",
+                encoding="utf-8",
+            ) as file:
+                reader = csv.reader(file, delimiter=";")
```
Is `csv.reader` more performant than rolling our own? It's just some hex digits from the starts of lines, right? I'd expect parsing the entire lines to take a little bit longer.

l4zygreed replied:
I thinks it's better for enduser if we don't do any parsing at all, doen't it? 🤔
I've replaced csv with python file and added instruction from which file (url) it's maken from, just in case. I think take first column from file and edit it into a python list is not a big deal for any other dev. (But I believe it will not be altered, at least for nearest few years.)

Ethsan replied:
Agreed, that probably the best solution. And that what seems to do kitten icat too (but in go).

toonn commented on ranger/ext/img_display.py:
```diff
@@ -792,6 +806,23 @@ def _late_init(self):
+        except CalledProcessError:
+            tmux_allow_passthrough = b'off'
+        if tmux_allow_passthrough == b'off':
+            raise ImageDisplayError("allow-passthrough no set in Tmux!")
```
```suggestion
            raise ImageDisplayError("allow-passthrough is not set in Tmux!")
```

toonn commented on ranger/ext/img_display.py:
```diff
@@ -898,6 +934,57 @@ def _format_cmd_str(self, cmd, payload=None, max_slice_len=2048):
+                # convert hex code to int and then to unicode char
+                self.diacritics = [chr(int(hex, 16)) for hex in hex_list]
+        except Exception as err:
+            raise ImageDisplayError("Error while loading diacritics" + str(err))
```
```suggestion
            raise ImageDisplayError("Error while loading diacritics: " + str(err))
```

toonn commented on ranger/ext/img_display.py:
```diff
@@ -691,9 +695,14 @@ class KittyImageDisplayer(ImageDisplayer, FileManagerAware):
+        https://sw.kovidgoyal.net/kitty/graphics-protocol/#display-images-on-screen"""
     protocol_start = b'\x1b_G'
     protocol_end = b'\x1b\\'
+    answer_start = protocol_start
+    answer_end = protocol_end
```
This part is honestly a bit confusing. Why are we initializing two sets of these? Do we need `answer_start` after changing `protocol_start` while also not being able to use it instead of `protocol_start` before changing it?

l4zygreed replied:
Because we need to send messages with escape sequence, but receive without it? Code doesn't have single "send code" function, where we can alter protocol, so it's easier to alter variables. Maybe we can refactor code and unify code sending? 🤔

toonn commented on ranger/ext/img_display.py:
```diff
@@ -720,6 +729,11 @@ def __init__(self):
         self.temp_file_dir = None  # Only used when streaming is not an option
 
     def _late_init(self):
+        # check if we inside tmux
+        # if true then we use the tmux escape sequence
```
```suggestion
        # Use Tmux escape sequence if appropriate.
```

toonn commented on ranger/ext/img_display.py:
```diff
@@ -792,6 +806,23 @@ def _late_init(self):
         self.needs_late_init = False
 
+    def _handle_init_tmux(self):
+        # check if tmux allows passthrough
+        # use tmux escape sequence if it's enabled
+        # throw error if it's disabled
```
```suggestion
        # Check whether allow-passthrough is enabled
```

toonn commented on ranger/ext/img_display.py:
```diff
@@ -857,9 +891,11 @@ def draw(self, path, start_x, start_y, width, height):
         # catch kitty answer before the escape codes corrupt the console
         resp = b''
-        while resp[-2:] != self.protocol_end:
+        while resp[-2:] != self.answer_end:
```
Why not use `endswith` here too?

l4zygreed replied:
replaced

toonn reviewed (commented):
FWIW in the future I'd prefer if people tried to PR to the PR branch of the original PR author, possibly linking to the downstream PR in our PR comments. It's inconvenient to keep track of things if people keep opening new PRs that are really the same thing. Rebasing on master is usually quite trivial so you can rest assured that that is *not* what holds back a PR from getting merged.

In this case Ethsan seems perfectly happy with the situation but it can come across as trying to take credit for someone else's work. All of this is easily avoided by opening the PR against their PR branch.

l4zygreed committed:
Update comments and error messages to requested

l4zygreed committed:
Update diacritics path to correct one

l4zygreed committed:
Replace slice comparison with endswith

l4zygreed committed:
Replace diacritics csv parsing with python code

l4zygreed committed:
Replace 033 code with x1b

In some apps code 027 should be used instead of 033, but x1b is
universal value

l4zygreed committed:
Do not change response_end if launched inside vim/nvim terminal

l4zygreed commented:
Sorry, been busy. (War and busy work - great combination, lol 🤣) 
Thought Ethsan would add the fixes. But I can work on PR a little bit now.

Ethsan commented:
I wanted too but I was busy and then completely forgot the PR. Sorry about that

Ethsan commented on ranger/ext/img_display.py:
```diff
@@ -3,6 +3,7 @@
 # Author: Emanuel Guevel, 2013
 # Author: Delisa Mason, 2015
 
+# pylint: disable=too-many-lines
```
I was thinking about may be splitting up the file like ext/vcs. But it's probably outside the scope of this PR

eMPee584 commented on ranger/ext/img_display.py:
```diff
@@ -805,6 +813,23 @@ def _late_init(self):
+        # throw error if it's disabled
+        try:
+            tmux_allow_passthrough = check_output(
+                ['tmux', 'show', '-Apv', 'allow-passthrough']).strip()
```
`tmux  show` is the shorthand for `tmux show-options`, I think it's better to use the long form because the meaning is more obvious..

l4zygreed committed:
Replace tmux show with tmux show-options

l4zygreed committed:
refactor: unify kitty write and read

Wrote protocol wrap, read, and write functions for kitty class. Also
replaced string concatenation with formatting to decrease a little
amount of operations and memory allocations.

l4zygreed commented:
@toonn Hello, can you review new commits? 🙃

l4zygreed committed:
Add KeyboardInterrupt to KIP read loop

toonn commented on ranger/ext/img_display.py:
```diff
@@ -719,7 +727,55 @@ def __init__(self):
+                self.tmux_protocol_start,
+                self.protocol_start,
+                msg,
+                self.tmux_protocol_end,
```
I *think* I see what you're doing with `tmux_protocol_{start,end}` but that seems confusing honestly, because of the interleaving.

And couldn't the `msg` contain `\x1b`? I think being explicit about doubling the `\x1b`s would be more straightforward code.

l4zygreed replied:
messages containing `\x1b` doesn't use wrap function.

but updated anyway, now should be less confusing.

toonn replied:
> messages containing `\x1b` doesn't use wrap function.

`wrapped` includes `protocol_{start,end}` so there's at least those `b'\x1b'`s, no?

l4zygreed replied:
in new code, yes. Messages which require wrapping are wrapped in single x1b's (protocol_start/end), if inside tmux - additionally wrapped into tmux_protocol_start/end with original x1b's doubled.

toonn commented on ranger/ext/img_display.py:
```diff
@@ -719,7 +727,55 @@ def __init__(self):
+        # Wrap bite string into protocol.
+        # With tmux passthrough if inside tmux.
+        if self.is_tmux:
+            return b'%s%s%s%s%s' % (
```
Let's use `"".format()` syntax, `%` is really old.

l4zygreed replied:
Bytes doesn't have `format` attribute, so this syntax AFAIK is the only formatting syntax for them.

Workarounds are:
* possible double conversion - format as string and then `encode` into bytes
* do something like `b''.join(protocal_start, msg, protocol_end)`

toonn replied:
My bad, %-formatting is fine then.

toonn reviewed (commented):
Some minor things I wanted to ask, let me know what you think.

l4zygreed committed:
fix typo in function name

l4zygreed committed:
move KITTY_PLACEHOLDER to diacritics file

l4zygreed committed:
update tmux wrapping in kitty protocol

Jagholin commented on ranger/ext/img_display.py:
```diff
@@ -727,8 +727,8 @@ def __init__(self):
 
-    def _wrap_bite_string(self, msg):
-        # Wrap bite string into protocol.
+    def _wrap_bytes_string(self, msg):
```
I think it's byte_string :)

l4zygreed replied:
another_one_bites_the_dust 🤔

toonn reviewed (approved):
I do think it's clearer this way. Looks ready now. Thank you for the contribution, previews working inside Tmux is awesome! :rocket: :tada:
